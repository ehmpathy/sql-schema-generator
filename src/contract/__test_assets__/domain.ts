import { Event } from '../../domain';
import { Entity, prop, ValueObject } from '../module';

const photo = new ValueObject({
  name: 'photo',
  properties: {
    url: prop.VARCHAR(255),
  },
});
const user = new Entity({
  name: 'user',
  properties: {
    phone_number: prop.CHAR(10), // only us numbers allowed
    first_name: prop.VARCHAR(255),
    last_name: prop.VARCHAR(255),
    avatar_id: {
      ...prop.REFERENCES(photo),
      updatable: true,
    },
  },
  unique: ['phone_number'], // users are identified by their phone number
});
const host = new Entity({
  name: 'host',
  properties: {
    tax_id: prop.VARCHAR(255),
    owner_id: prop.REFERENCES(user),
    contact_id: {
      ...prop.REFERENCES(user),
      updatable: true,
    },
  },
  unique: ['tax_id'], // hosts are identified by their tax ids
});
const home = new Entity({
  name: 'home',
  properties: {
    name: prop.VARCHAR(255),
    host_ids: prop.ARRAY_OF(prop.REFERENCES(host)), // one home may have more than one host (for some reason in this example... just go with it)
    built: prop.TIMESTAMPTZ(),
    bedrooms: prop.INT(),
    bathrooms: prop.INT(),
    photo_ids: {
      ...prop.ARRAY_OF(prop.REFERENCES(photo)),
      updatable: true, // the photos of a home change over time
    },
  },
  unique: ['name', 'host_ids'],
});
const welcomedHomeEvent = new Event({
  name: 'welcomed_home_event',
  properties: {
    occurred_at: prop.DATE(),
    home_id: prop.REFERENCES(home),
    user_id: prop.REFERENCES(user),
  },
  unique: ['occurred_at', 'home_id'], // only one user can be welcomed at a time
});
const message = new Entity({
  name: 'message',
  properties: {
    about_home_id: prop.REFERENCES(home),
    from_user_id: prop.REFERENCES(user),
    text: prop.VARCHAR(), // the text they sent as part of the message
    image_uuids: prop.ARRAY_OF(prop.UUID()), // the images they sent as part of the message (e.g., references a table in a different db by uuid)
  },
  unique: ['uuid'],
});

export const generateSqlSchemasFor = [
  photo,
  user,
  host,
  home,
  welcomedHomeEvent,
  message,
];
