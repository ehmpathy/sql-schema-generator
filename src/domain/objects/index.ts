export { DataType } from './DataType';
export { Property } from './Property';
export { Entity, Properties } from './Entity';
export { ValueObject } from './ValueObject';
