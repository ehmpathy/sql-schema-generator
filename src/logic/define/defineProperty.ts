/*
  purpose: provide convenient tools to define types
*/
import { DataType, DataTypeName, Entity, Property } from '../../types';

/**
 * TINYINT: requires 1 byte of storage. Unsigned range [-128, 127]. Signed range [0, 255].
 *
 * https://dev.mysql.com/doc/refman/5.7/en/integer-types.html
 */
export const TINYINT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.TINYINT,
      precision: 4,
    }),
  });

/**
 * SMALLINT: requires 2 bytes of storage. Unsigned range [-32,768, 32,768]. Signed range [0, 65,535].
 *
 * https://dev.mysql.com/doc/refman/5.7/en/integer-types.html
 */
export const SMALLINT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.SMALLINT,
      precision: 6,
    }),
  });

/**
 * MEDIUMINT: requires 3 bytes of storage. Unsigned range [-8,388,608, 8,388,607]. Signed range [0, 16,777,215].
 *
 * https://dev.mysql.com/doc/refman/5.7/en/integer-types.html
 */
export const MEDIUMINT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.MEDIUMINT,
      precision: 9,
    }),
  });

/**
 * INT: requires 4 bytes of storage. Unsigned range [-2,147,483,648, 2,147,483,647]. Signed range [0, 4,294,967,295].
 *
 * https://dev.mysql.com/doc/refman/5.7/en/integer-types.html
 */
export const INT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.INT,
      precision: 11,
    }),
  });

/**
 * BIGINT: requires 8 bytes of storage. Unsigned range [-2^63, 2^63-1]. Signed range [0, 2^64-1].
 *
 * https://dev.mysql.com/doc/refman/5.7/en/integer-types.html
 */
export const BIGINT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.BIGINT,
      precision: 20,
    }),
  });

/**
 * The DECIMAL type stores exact numeric data values. These types are used when it is important to preserve exact precision, for example with monetary data.
 *
 * Precision represents the number of significant digits that are stored in total. Scale represents the number of digits following the decimal point.
 *
 * Example: the range of DECIMAL(5,2) is [-999.99, 999.99], since 5 is the precision and 2 is the scale.
 *
 * https://dev.mysql.com/doc/refman/5.7/en/fixed-point-types.html
 */
export const DECIMAL = (precision: number, scale: number) =>
  new Property({
    type: new DataType({
      name: DataTypeName.DECIMAL,
      precision,
      scale,
    }),
  });

/**
 * The FLOAT type represents approximate numeric data values.
 *
 * FLOAT requires 4 bytes.
 *
 * https://dev.mysql.com/doc/refman/5.7/en/floating-point-types.html
 */
export const FLOAT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.FLOAT,
    }),
  });

/**
 * The DOUBLE type represents approximate numeric data values.
 *
 * DOUBLE requires 8 bytes.
 *
 * https://dev.mysql.com/doc/refman/5.7/en/floating-point-types.html
 */
export const DOUBLE = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.DOUBLE,
    }),
  });

/**
 * The BIT data type is used to store bit values. A type of BIT(M) enables storage of M-bit values. M can range from 1 to 64.
 *
 * If you assign a value to a BIT(M) column that is less than M bits long, the value is padded on the left with zeros. For example, assigning a value of b'101' to a BIT(6) column is, in effect, the same as assigning b'000101'.
 *
 * https://dev.mysql.com/doc/refman/5.7/en/bit-type.html
 */
export const BIT = (precision: number) =>
  new Property({
    type: new DataType({
      name: DataTypeName.BIT,
      precision,
    }),
  });

/**
 * UUID is an alias for CHAR(36)
 */
export const UUID = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.CHAR,
      precision: 36,
    }),
    // TODO: add check for proper uuid format
  });

/**
 * An ENUM is a string object with a value chosen from a list of permitted values that are enumerated explicitly in the column specification at table creation time.
 *
 * WARNING: An enumeration value can also be the empty string ('') or NULL under certain circumstances, per MySQL documentation (!). Only if strict SQL mode is enabled, attempts to insert invalid ENUM values result in an error.
 *
 * https://dev.mysql.com/doc/refman/5.7/en/enum.html
 */
export const ENUM = (values: string[]) =>
  new Property({
    type: new DataType({
      name: DataTypeName.ENUM,
      values,
    }),
  });

/**
 * CHAR(n) requires n bytes. Precision defines the maximum number of characters you want to store. Range of precision: [0, 255].
 *
 * When CHAR values are stored, they are right-padded with spaces to the specified length. When CHAR values are retrieved, trailing spaces are removed unless the PAD_CHAR_TO_FULL_LENGTH SQL mode is enabled.
 *
 * WARNING: If strict SQL mode is not enabled and you assign a value to a CHAR column that exceeds the column's maximum length, the value is truncated to fit and a warning is generated.
 *
 * https://dev.mysql.com/doc/refman/5.7/en/char.html
 */
export const CHAR = (precision: number) =>
  new Property({
    type: new DataType({
      name: DataTypeName.VARCHAR,
      precision,
    }),
  });

/**
 * VARCHAR requires a variable storage size. Precision defines the maximum number of characters you want to store. Range of precision: [0, 65,535].
 *
 * VARCHAR values are stored as a 1-byte or 2-byte length prefix plus data. The length prefix indicates the number of bytes in the value. A column uses one length byte if values require no more than 255 bytes, two length bytes if values may require more than 255 bytes.
 *
 * WARNING: If strict SQL mode is not enabled and you assign a value to a VARCHAR column that exceeds the column's maximum length, the value is truncated to fit and a warning is generated.
 *
 * https://dev.mysql.com/doc/refman/5.7/en/char.html
 */
export const VARCHAR = (precision: number) =>
  new Property({
    type: new DataType({
      name: DataTypeName.VARCHAR,
      precision,
    }),
  });

/**
 * TINYTEXT requires L + 1 bytes, where L < 2^8. (2^8 = 256)
 *
 * https://dev.mysql.com/doc/refman/5.7/en/blob.html
 */
export const TINYTEXT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.TINYTEXT,
    }),
  });

/**
 * TEXT requires L + 2 bytes, where L < 2^16. (2^16 = 65,536)
 *
 * https://dev.mysql.com/doc/refman/5.7/en/blob.html
 */
export const TEXT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.TEXT,
    }),
  });

/**
 * TEXT requires L + 3 bytes, where L < 2^24. (2^32 = 16,777,216)
 *
 * https://dev.mysql.com/doc/refman/5.7/en/blob.html
 */
export const MEDIUMTEXT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.MEDIUMTEXT,
    }),
  });

/**
 * TEXT requires L + 4 bytes, where L < 2^32. (2^32 = 4,294,967,296)
 *
 * https://dev.mysql.com/doc/refman/5.7/en/blob.html
 */
export const LONGTEXT = () =>
  new Property({
    type: new DataType({
      name: DataTypeName.LONGTEXT,
    }),
  });

/**
 * https://dev.mysql.com/doc/refman/5.7/en/binary-varbinary.html
 */
export const BINARY = (precision: number) =>
  new Property({
    type: new DataType({
      name: DataTypeName.BINARY,
      precision,
    }),
  });

/**
 * The DATETIME type is used for values that contain both date and time parts. MySQL retrieves and displays DATETIME values in 'YYYY-MM-DD hh:mm:ss' format. The supported range is '1000-01-01 00:00:00' to '9999-12-31 23:59:59'.
 *
 * DATETIME requires 8 bytes of storage.
 *
 * https://dev.mysql.com/doc/refman/5.7/en/datetime.html
 * https://dev.mysql.com/doc/refman/5.7/en/storage-requirements.html
 */
export const DATETIME = (precision: number) =>
  new Property({
    type: new DataType({
      name: DataTypeName.DATETIME,
      precision,
    }),
  });

/**
 * REFERENCES creates a column which has a Foreign Key constraint to the `entity`.
 */
export const REFERENCES = (entity: Entity) =>
  new Property({
    ...BIGINT(), // pk type is always a bigint
    references: entity.name, // name of entity
  });
