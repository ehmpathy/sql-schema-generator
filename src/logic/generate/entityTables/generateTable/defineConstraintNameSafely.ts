import { toHashShake256Sync } from '../../../../__nonpublished_modules__/hash-fns/toHashShake256Sync';

const POSTGRES_RESOURCE_NAME_LEN_LIMIT = 63;

/**
 * .what = determines an intuitive yet safe constraint name
 * .why = postgres has a 63 char resource name limit
 * .how = include a hash if the ideal name would be too long
 */
export const defineConstraintNameSafely = (input: {
  tableName: string;
  constraintName: string;
}): string => {
  const idealConstraintName = [input.tableName, input.constraintName].join('_');
  if (idealConstraintName.length <= POSTGRES_RESOURCE_NAME_LEN_LIMIT)
    return idealConstraintName;
  const hash = toHashShake256Sync(input.tableName, 4);
  const tableNameLimit =
    POSTGRES_RESOURCE_NAME_LEN_LIMIT -
    hash.length -
    input.constraintName.length -
    2;
  const safeConstraintName = [
    input.tableName.slice(0, tableNameLimit),
    hash,
    input.constraintName,
  ].join('_');
  return safeConstraintName;
};
