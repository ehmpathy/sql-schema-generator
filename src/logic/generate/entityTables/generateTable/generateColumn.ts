import { DataTypeName, Property } from '../../../../types';
import { extractMysqlTypeDefinitionFromProperty } from '../../utils/extractMysqlTypeDefinitionFromProperty';

/*
  // TODO: generalize w/ adapter pattern to other languages
*/
const textTypes = [
  DataTypeName.CHAR,
  DataTypeName.VARCHAR,
  DataTypeName.TINYTEXT,
  DataTypeName.TEXT,
  DataTypeName.MEDIUMTEXT,
  DataTypeName.LONGTEXT,
  DataTypeName.ENUM,
];
export const generateColumn = ({ columnName, property }: { columnName: string, property: Property }) => {

  // define default over-rides
  let defaultNullRelation = (property.nullable) ? 'DEFAULT NULL' : ''; // if nullable, then the "default" default is "null"; else, nothing
  if (property.type.name === DataTypeName.TEXT) defaultNullRelation = ''; // if its varchar, then even if nullable don't show default as null; // TODO: discover why SHOW CREATE has this exception

  // define the full column definition
  const modifiers = [
    `\`${columnName}\``,

    extractMysqlTypeDefinitionFromProperty({ property }), // e.g., property => bigint(20)

    (textTypes.includes(property.type.name)) ? 'COLLATE utf8_bin' : '', // coalate if its a string type; collate to Case Sensitive by default

    (property.nullable) ? '' : 'NOT NULL', // if not nullable, then NOT NULL

    (property.default) ? `DEFAULT ${property.default}` : defaultNullRelation, // if default is set, use it; otherwise, no default

    (property.comment) ? `COMMENT '${property.comment}'` : '', // empty string if comment not defined

    (columnName === 'id') ? 'AUTO_INCREMENT' : '',
  ];
  return modifiers
    .filter(modifier => !!modifier) // filter out null
    .join(' '); // merge into column
};
