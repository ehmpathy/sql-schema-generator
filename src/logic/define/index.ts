import * as prop from './defineProperty';

export { prop };
