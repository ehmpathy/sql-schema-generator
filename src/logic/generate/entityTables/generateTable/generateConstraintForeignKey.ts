import { Property } from '../../../../domain';
import { defineConstraintNameSafely } from './defineConstraintNameSafely';

export const generateConstraintForeignKey = ({
  index,
  tableName,
  columnName,
  property,
}: {
  index: number;
  tableName: string;
  columnName: string;
  property: Property;
}) => {
  const constraintName = `${tableName}`;
  const constraintSql = `
CONSTRAINT ${defineConstraintNameSafely({
    tableName,
    constraintName: `fk${index}`,
  })} FOREIGN KEY (${columnName}) REFERENCES ${property.references!} (id)
  `.trim();

  const indexSql = `
CREATE INDEX ${defineConstraintNameSafely({
    tableName,
    constraintName: `fk${index}_ix`,
  })} ON ${tableName} USING btree (${columnName});
  `.trim();

  return {
    index: indexSql,
    constraint: constraintSql,
  };
};
