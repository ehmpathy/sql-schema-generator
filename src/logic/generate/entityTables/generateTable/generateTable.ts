import { Property } from '../../../../domain';
import { generateColumn } from './generateColumn';
import { generateConstraintForeignKey } from './generateConstraintForeignKey';

export const generateTable = ({
  tableName,
  properties,
  unique,
}: {
  tableName: string;
  properties: { [index: string]: Property };
  unique: string[];
}) => {
  // 0. validate input
  if (unique.length === 0) {
    throw new Error('must have atleast one unique property; otherwise, idempotency cant be enforced');
  }

  // define sql per column
  const columnSqls = Object.entries(properties).map((entry) =>
    generateColumn({ columnName: entry[0], property: entry[1] }),
  );

  // define primary key
  const primaryKeySql = `CONSTRAINT ${tableName}_pk PRIMARY KEY (id)`;

  // define unique index
  const uniqueConstraintSql = `CONSTRAINT ${tableName}_ux1 UNIQUE (${unique.join(', ')})`; // unique key definition; required since it is required for idempotency

  // define foreign keys
  const foreignKeySqls = Object.entries(properties)
    .filter((entry) => !!entry[1].references)
    .map((entry, index) =>
      generateConstraintForeignKey({ index, columnName: entry[0], tableName, property: entry[1] }),
    );
  const foreignKeyIndexSqls = foreignKeySqls.map((sqls) => sqls.index);
  const foreignKeyConstraintSqls = foreignKeySqls.map((sqls) => sqls.constraint);

  // define check constraints
  const checkConstraintSqls = Object.entries(properties)
    .filter((entry) => !!entry[1].check)
    .map((entry) => {
      return `CONSTRAINT ${tableName}_${entry[0]}_check CHECK ${entry[1].check!.replace(/\$COLUMN_NAME/g, entry[0])}`;
    });

  // 2. define the lines of content of the table
  const contents = [
    ...columnSqls, // all of the columns (one per property)
    primaryKeySql,
    uniqueConstraintSql,
    ...foreignKeyConstraintSqls, // constraints defined for FKs
    ...checkConstraintSqls, // constraints defined for checks
  ];

  // define the indexes we're creating
  const indexes = [
    // NOTE: pk index is auto created
    // NOTE: ux index is auto created
    ...foreignKeyIndexSqls, // indexes defined for FKs
  ];

  // 2. generate the create table sql
  const sql = `
CREATE TABLE ${tableName} (
  ${contents.join(',\n  ')}
);
${indexes.join('\n')}
  `.trim();

  // 3. return the sql
  return sql;
};
