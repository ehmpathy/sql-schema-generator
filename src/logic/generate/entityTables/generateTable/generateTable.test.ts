import { DataType, DataTypeName, Property } from '../../../../domain';
import * as prop from '../../../define/defineProperty';
import { generateColumn } from './generateColumn';
import { generateConstraintForeignKey } from './generateConstraintForeignKey';
import { generateTable } from './generateTable';

jest.mock('./generateColumn');
const generateColumnMock = generateColumn as jest.Mock;
generateColumnMock.mockReturnValue('__COLUMN__');

jest.mock('./generateConstraintForeignKey');
const generateConstraintForeignKeyMock =
  generateConstraintForeignKey as jest.Mock;
generateConstraintForeignKeyMock.mockReturnValue({
  constraint: '__FOREIGN_KEY_CONSTRAINT__',
  index: '__FOREIGN_KEY_INDEX__',
});

describe('generateTableConstraint', () => {
  beforeEach(() => jest.clearAllMocks());
  const userIdProperty = new Property({
    type: new DataType({
      name: DataTypeName.INT,
    }),
    references: 'user',
  });
  const stationIdProperty = new Property({
    type: new DataType({
      name: DataTypeName.INT,
    }),
    references: 'train_station',
  });
  const statusProperty = prop.ENUM(['happy', 'meh', 'sad']);
  it('should define the table with the correct name', () => {
    const sql = generateTable({
      tableName: 'message',
      properties: { user_id: userIdProperty },
      unique: ['user_id'],
    });
    expect(sql).toContain('CREATE TABLE message');
  });
  it('should call generateColumn for each property', () => {
    generateTable({
      tableName: 'message',
      properties: { user_id: userIdProperty },
      unique: ['user_id'],
    });
    expect(generateColumnMock.mock.calls.length).toEqual(1);
    expect(generateColumnMock.mock.calls[0][0]).toMatchObject({
      property: userIdProperty,
    });
  });
  it('should call generateConstraintForeignKey for each references property', () => {
    generateTable({
      tableName: 'message',
      properties: { user_id: userIdProperty },
      unique: ['user_id'],
    });
    expect(generateConstraintForeignKeyMock.mock.calls.length).toEqual(1);
    expect(generateConstraintForeignKeyMock.mock.calls[0][0]).toMatchObject({
      property: userIdProperty,
    });
  });
  it('should define the columns', () => {
    const sql = generateTable({
      tableName: 'message',
      properties: { user_id: userIdProperty },
      unique: ['user_id'],
    });
    expect(sql).toContain('__COLUMN__,');
  });
  it('should define id as the primary key', () => {
    const sql = generateTable({
      tableName: 'message',
      properties: { user_id: userIdProperty },
      unique: ['user_id'],
    });
    expect(sql).toContain('CONSTRAINT message_pk PRIMARY KEY (id),');
  });
  it('should define the unique key', () => {
    const sql = generateTable({
      tableName: 'message',
      properties: { user_id: userIdProperty },
      unique: ['user_id'],
    });
    expect(sql).toContain('CONSTRAINT message_ux1 UNIQUE (user_id),');
  });
  it('should define the foreign key constraints', () => {
    const sql = generateTable({
      tableName: 'message',
      properties: { user_id: userIdProperty },
      unique: ['user_id'],
    });
    expect(sql).toContain('__FOREIGN_KEY_INDEX__');
    expect(sql).toContain('__FOREIGN_KEY_CONSTRAINT__');
  });
  it('should not have empty lines if no foreign keys are defined', () => {
    const sql = generateTable({
      tableName: 'message',
      properties: { status: statusProperty },
      unique: ['status'],
    });
    expect(sql).not.toMatch(/^\s*,?$/m); // no lines should be empty or only contain spaces and a comma
  });
  it('should throw an error if no unique key columns are specified', () => {
    try {
      generateTable({
        tableName: 'message',
        properties: { status: statusProperty },
        unique: [],
      });
      throw new Error('should not reach here');
    } catch (error) {
      expect(error.message).toEqual(
        'must have atleast one unique property; otherwise, idempotency cant be enforced',
      );
    }
  });
  it('should safely define the constraint names for constraints that would otherwise have names longer than 64 chars', () => {
    const sql = generateTable({
      tableName:
        'async_task_predict_train_station_congestion_per_movement_event',
      properties: { status: statusProperty, stationId: stationIdProperty },
      unique: ['status'],
    });
    console.log(sql);
    expect(sql).toMatchSnapshot();
  });
});
