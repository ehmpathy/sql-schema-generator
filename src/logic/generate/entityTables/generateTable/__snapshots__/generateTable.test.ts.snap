// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTableConstraint should safely define the constraint names for constraints that would otherwise have names longer than 64 chars 1`] = `
"CREATE TABLE async_task_predict_train_station_congestion_per_movement_event (
  __COLUMN__,
  __COLUMN__,
  CONSTRAINT async_task_predict_train_station_congestion_per_mov_50cc24a7_pk PRIMARY KEY (id),
  CONSTRAINT async_task_predict_train_station_congestion_per_mo_50cc24a7_ux1 UNIQUE (status),
  __FOREIGN_KEY_CONSTRAINT__,
  CONSTRAINT async_task_predict_train_station_congesti_50cc24a7_status_check CHECK (status IN ('happy', 'meh', 'sad'))
);
__FOREIGN_KEY_INDEX__"
`;
